AK/JsonArraySerializer.h

```diff
@@ -1,139 +1,233 @@
 /*
  * Copyright (c) 2019-2020, Sergey Bugaev <bugaevc@serenityos.org>
+ * Copyright (c) 2022, Idan Horowitz <idan.horowitz@serenityos.org>
  *
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
 #pragma once
 
-#include <AK/JsonValue.h>
+#include <AK/Error.h>
+#include <AK/Try.h>
+
+#ifndef KERNEL
+#    include <AK/JsonValue.h>
+#endif
 
 namespace AK {
 
 template<typename Builder>
+inline constexpr bool IsLegacyBuilder = requires(Builder builder) { builder.try_append('\0'); };
+
+template<typename Builder = void>
 class JsonObjectSerializer;
 
-template<typename Builder>
+template<typename Builder = void>
 class JsonArraySerializer {
 public:
-    explicit JsonArraySerializer(Builder& builder)
-        : m_builder(builder)
+    static ErrorOr<JsonArraySerializer> try_create(Builder& builder)
+    {
+        if constexpr (IsLegacyBuilder<Builder>)
+            TRY(builder.try_append('['));
+        else
+            TRY(builder.append('['));
+        return JsonArraySerializer { builder };
+    }
+
+    JsonArraySerializer(JsonArraySerializer&& other)
+        : m_builder(other.m_builder)
+        , m_empty(other.m_empty)
+        , m_finished(exchange(other.m_finished, true))
     {
-        (void)m_builder.append('[');
     }
 
     JsonArraySerializer(const JsonArraySerializer&) = delete;
-    JsonArraySerializer(JsonArraySerializer&&) = delete;
 
     ~JsonArraySerializer()
     {
-        if (!m_finished)
-            finish();
+        VERIFY(m_finished);
     }
 
 #ifndef KERNEL
-    void add(const JsonValue& value)
+    ErrorOr<void> add(const JsonValue& value)
     {
-        begin_item();
+        TRY(begin_item());
         value.serialize(m_builder);
+        return {};
     }
 #endif
 
-    void add(StringView value)
+    ErrorOr<void> add(StringView value)
     {
-        begin_item();
-        (void)m_builder.append('"');
-        (void)m_builder.append_escaped_for_json(value);
-        (void)m_builder.append('"');
+        TRY(begin_item());
+        if constexpr (IsLegacyBuilder<Builder>) {
+            TRY(m_builder.try_append('"'));
+            TRY(m_builder.try_append_escaped_for_json(value));
+            TRY(m_builder.try_append('"'));
+        } else {
+            TRY(m_builder.append('"'));
+            TRY(m_builder.append_escaped_for_json(value));
+            TRY(m_builder.append('"'));
+        }
+        return {};
```

Comment:
This seems like using exceptions, but with another extra step. 

Just saying.

Reply:
Please avoid leaving useless, unactionable comments on PRs.